### PR TITLE
Flutter inspector implemented in Flutter

### PR DESCRIPTION
**Pull request by jacob314**

### CODE REVIEW

**Comment by jacob314** [comment]
@terrylucas and @kenzieschmoll 
This CL will probably illustrate a lot of noob Flutter developer mistakes that you will want to avoid as you port DevTools code to Flutter.

**Comment by jacob314** [comment]
The inspector_controller has good test coverage. At some point we should add some integration tests that run the inspector_screen connected to a live flutter app capturing screenshots of the widget tree.

File: packages/devtools_app/test/flutter/inspector_screen_test.dart [hunk 1]
```diff
@@ -168,5 +170,9 @@ void main() {
       // TODO(jacobr): also verify that the service extension buttons look
       // visually disabled.
     });
+
+    // TODO(jacobr): add screenshot tests that connect to a test application
```
Comment (DaveShuckerow) [inline packages/devtools_app/test/flutter/inspector_screen_test.dart]: How would you feel about using pre-captured VM snapshots for this one?  That will save us the trouble of bootstrapping an integration test environment and these screenshot tests can be run as simple unit tests.

File: packages/devtools_app/lib/src/flutter/blocking_action_mixin.dart [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+///
+/// Use this mixin to simplifying showing buttons as disabled while blocking
+/// actions are being performed and otherwise avoid executing multiple actions
+/// at once.
+mixin BlockingActionMixin<T extends StatefulWidget> on State<T> {
+  /// Returns whether an action is in progress.
+  ///
+  /// Typically users should disable buttons and or other UI that should not
+  /// be interactive while the action is in progress.
+  bool get actionInProgress => _actionInProgress;
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/blocking_action_mixin.dart]: done

File: packages/devtools_app/lib/src/inspector/flutter/inspector_screen.dart [hunk 3]
```diff
@@ -25,42 +36,223 @@ class InspectorScreen extends Screen {
...
+        detailsTree: detailsTreeController,
+        inspectorService: inspectorService,
+        treeType: FlutterTreeType.widget,
+        onExpandCollapseSupported: _onExpandCollapseSupported,
+      );
+
+      // TODO(jacobr): update visibility based on whether the screen is visible.
+      // That will reduce memory usage on the device running a Flutter
+      // application when the inspector panel is not visible.
+      inspectorController.setVisibleToUser(true);
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_screen.dart]: good point. done.
added the extra plumbing from the inspectorTreeController to the inspectorController so that the inspectorController deactivates when the last inspectorTreeView deactivates.

File: packages/devtools_app/lib/src/inspector/flutter/inspector_screen.dart [hunk 4]
```diff
@@ -25,42 +36,223 @@ class InspectorScreen extends Screen {
...
+      if (!displayedWidgetTrackingNotice) {
+        // ignore: unawaited_futures
+        inspectorService.isWidgetCreationTracked().then((bool value) {
+          if (value) {
+            return;
+          }
+
+          displayedWidgetTrackingNotice = true;
+          // TODO(jacobr): implement showMessage.
+          // framework.showMessage(
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_screen.dart]: it is the semi-persistent dismissible nag header we'd like to add.

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+  final _InspectorTreeState inspectorTreeState;
+
+  InspectorTreeNode get node => row.node;
+  final InspectorTreeRow row;
+
+  @override
+  _InspectorTreeRowState createState() => _InspectorTreeRowState();
+}
+
+///
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: done

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+              ),
+            ],
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+class _InspectorTreeRowState extends State<_InspectorTreeRowWidget>
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: @pq it should be a lint if you ever accidentally add content between the State object and corresponding widget.

[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: done

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 7]
```diff
@@ -0,0 +1,492 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: done

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (pq) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: @jacob314: could you give me a concrete example?

File: packages/devtools_app/lib/src/inspector/flutter/inspector_screen.dart [hunk 8]
```diff
@@ -25,42 +36,223 @@ class InspectorScreen extends Screen {
...
+        connectionInProgress = false;
+      });
+    }
+
+    if (inspectorService == null) {
+      return;
+    }
+
+    setState(() {
+      summaryTreeController = InspectorTreeControllerFlutter();
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_screen.dart]: the inspectorController does. I've added the missing dispose call for it.

File: packages/devtools_app/lib/src/inspector/flutter/inspector_screen.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_screen.dart]: we are. the dispose method is clearing the client.

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]:
```dart
class FooWidget extends StatefulWidget {
  @override
  _FooWidgetState createState() => _FooWidgetState();
}

class OopsIDefinedAClassHere { // Lint this as I made it harder for a reviewer to see the State and Widget classes at the same time.
  
}

class _FooWidgetState extends State<FooWidget> {
  @override
  Widget build(BuildContext context) {
    return Container();
  }
}
```

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+    return maxOffset - viewportDimension;
+  }
+
+  @override
+  void didUpdateWidget(Widget oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _bindToController();
+  }
+
+  void _bindToController() {
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: it is disposed in the dispose method a couple lines lower.

File: packages/devtools_app/test/flutter/inspector_screen_test.dart (hunk 1 above)
Comment (jacob314) [reply packages/devtools_app/test/flutter/inspector_screen_test.dart]: at some point the inspector will be refactored enough to support running offline based on json snapshots describing the app. Until that is done, this is not as simple as you might expect.

Comment (jacob314) [reply packages/devtools_app/test/flutter/inspector_screen_test.dart]: you need a full flutter system and there is a lot of chat back on forth over the widget_inspector api to render.

**Comment by jacob314** [comment]
ptal

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+}
+
+abstract class InspectorControllerClient {
+  void onChanged();
+
+  void scrollToRect(Rect rect);
+}
+
+class _InspectorTreeState extends State<InspectorTree>
+    implements InspectorControllerClient {
```
Comment (DaveShuckerow) [inline packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: This is an interesting take on the pattern of defining listeners.  It's different from the standard approach of handing tearoffs to the controller call back when asking for a change, and it gave me pause at first. 

Semantically the approaches are the same.  Go for it.

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 11]
```diff
@@ -490,3 +362,144 @@ class InspectorTree extends StatefulWidget {
   @override
   State<InspectorTree> createState() => _InspectorTreeState();
 }
+
+final _defaultPaint = Paint()
+  ..color = defaultTreeLineColor
+  ..strokeWidth = chartLineStrokeWidth;
+
+class _RowPainter extends CustomPainter {
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: done. commented and removed some dead code.

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 12]
```diff
@@ -0,0 +1,505 @@
...
+              controller: _scrollControllerY,
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+class InspectorTree extends StatefulWidget {
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: done

File: packages/devtools_app/lib/src/flutter/inspector/diagnostics.dart [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+
+const bool _showRenderObjectPropertiesAsLinks = false;
+
+/// Presents the content of a single [RemoteDiagnosticsNode].
+///
+/// Use this class any time you want to display a single [RemoteDiagnosticsNode]
+/// in debugging UI whether you are displaying the node in the [InspectorTree]
+/// in console output, or a debugger.
+/// See also:
+/// * [InspectorTree], which uses this class to display each node in the in
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/inspector/diagnostics.dart]: typo: remove extra "in" at end of line

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 14]
```diff
@@ -0,0 +1,505 @@
...
+    _bindToController();
+  }
+
+  void _bindToController() {
+    controller?.client = this;
+  }
+
+  @override
+  void onChanged() {
+    setState(() => {});
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: done

[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: actually this code technically returns a dart Set or Map literal so the lint will be a bit tricky to write without having false positives.

File: packages/devtools_app/lib/src/flutter/inspector/diagnostics.dart [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+        output.add(Text(match.group(1), style: textStyle));
+        if (match.group(2).isNotEmpty) {
+          output.add(Text(
+            match.group(2),
+            style: inspector_text_styles.unimportant,
+          ));
+        }
+        return;
+      }
+    } else if (diagnostic.type == 'ErrorDescription') {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/inspector/diagnostics.dart]: pull 'ErrorDescription' out into a const

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart (hunk 10 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: Can you give me a concrete example of the pattern you are suggesting? Happy to also just see a CL refactoring this to the traditional pattern. I didn't mean to do something non-standard here.

File: packages/devtools_app/lib/src/flutter/inspector/diagnostics.dart [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+
+    if (icon != null) {
+      children.add(_toFlutterIcon(icon));
+    }
+    final String name = diagnostic.name;
+    TextStyle textStyle = textStyleForLevel(diagnostic.level);
+    if (diagnostic.isProperty) {
+      // Display of inline properties.
+      final String propertyType = diagnostic.propertyType;
+      final Map<String, Object> properties = diagnostic.valuePropertiesJson;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/inspector/diagnostics.dart]: can we remove some of these types?

File: packages/devtools_app/lib/src/flutter/inspector/diagnostics.dart [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+
+      String description = diagnostic.description;
+      if (propertyType != null && properties != null) {
+        switch (propertyType) {
+          case 'Color':
+            {
+              final int alpha = JsonUtils.getIntMember(properties, 'alpha');
+              final int red = JsonUtils.getIntMember(properties, 'red');
+              final int green = JsonUtils.getIntMember(properties, 'green');
+              final int blue = JsonUtils.getIntMember(properties, 'blue');
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/inspector/diagnostics.dart]: same here and elsewhere: remove types?

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart (hunk 10 above)
Comment (DaveShuckerow) [reply packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]:
```dart
controller = Controller(
  onSomething: _doSomething,
  onSomethingElse: _doSomethingElse,
);
```

or, for a longer-lived controller:

```dart
controller.addListener(
  onSomething: _doSomething,
  onSomethingElse: _doSomethingElse,
);
```

But your approach here with a defined interface is equivalent.

So long as we are initState()ing and dispose()ing properly, there's nothing to be concerned about.

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 18]
```diff
@@ -0,0 +1,511 @@
...
+                    color: backgroundColor,
+                  ),
+                  child: InkWell(
+                    onTap: () {
+                      controller.onSelectRow(row);
+                    },
+                    child: Container(
+                        height: rowHeight,
+                        padding: const EdgeInsets.symmetric(horizontal: 4.0),
+                        child: DiagnosticsNodeDescription(node.diagnostic)),
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: nit: trailing comma

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 19]
```diff
@@ -0,0 +1,511 @@
...
+              node.showExpandCollapse
+                  ? InkWell(
+                      onTap: onToggle,
+                      child: RotationTransition(
+                        turns: expandAnimation,
+                        child: Icon(
+                          Icons.expand_more,
+                          size: 16.0,
+                        ),
+                      ))
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: nit: comma

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 20]
```diff
@@ -0,0 +1,511 @@
...
+    return false;
+  }
+}
+
+/// Widget defining the contents of a single row in the InspectorTree.
+///
+/// This class defines the scaffolding around the rendering of the actual
+/// content of a [RemoteDiagnosticsNode] provided by
+/// [DiagnosticsNodeDescription] to provide a tree implementation with lines
+/// drawn between parent and child nodes when nodes have multiple children.
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: nit: this documentation is a bit hard to parse. Breaking this up into multiple sentences would help.

File: packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart [hunk 21]
```diff
@@ -0,0 +1,511 @@
...
+  double get maxRowIndent {
+    if (lastContentWidth == null) {
+      double maxIndent = 0;
+      for (int i = 0; i < numRows; i++) {
+        final row = getCachedRow(i);
+        if (row != null) {
+          maxIndent = max(maxIndent, getDepthIndent(row.depth));
+        }
+      }
+      lastContentWidth = maxIndent + maxIndent;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart]: this confuses me. Why is lastContentWidth twice the width of the indent?
